### DISCUSSION TITLE
Writing a Formulae for KMonad which depends on a Cask during runtime

### DISCUSSION THREAD

**Discussion by niklaas** [Homebrew/discussions#5015] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.1-30-gf3c687e
ORIGIN: https://github.com/Homebrew/brew
HEAD: f3c687e4728d5a678e9ebd63a9bf95e7cc7a0ac0
Last commit: 2 days ago
Core tap HEAD: 9faaea5efe8a0a51ff951ad4dc753f68952b3aa7
Core tap last commit: 6 hours ago
Core tap JSON: 30 Dec 16:22 UTC
Core cask tap HEAD: a92a35f1c6c53ad800706f28ee64568620a505ef
Core cask tap last commit: 5 hours ago
Core cask tap JSON: 30 Dec 16:22 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 16
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 16-core 64-bit arm_palma
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.1.1-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: A newer Command Line Tools release is available.
Update them from Software Update in System Settings.

If that doesn't show you any updates, run:
  sudo rm -rf /Library/Developer/CommandLineTools
  sudo xcode-select --install

Alternatively, manually download them from:
  https://developer.apple.com/download/all/.
You should download the Command Line Tools for Xcode 15.1.


Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
  ?? Formula/k/kmonad.rb

You have uncommitted modifications to Homebrew/homebrew-cask.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-cask && git stash -u && git clean -d -f

Uncommitted files:
  ?? Casks/k/karabiner-driverkit.rb
```

### Description of issue

Hi! I'm working on a Formulae for [KMonad](https://github.com/kmonad/kmonad). They have [pretty detailed installation instructions](https://github.com/kmonad/kmonad/blob/master/doc/installation.md#macos) and I am able to build it locally on Mac OS. However, I am wondering about how to write the Formulae given that KMonad depends on a Cask during runtime.

KMonad depends on "system extensions". When building, you can decide between two different system extensions: either `kext` or `dext`. `kext` is for older Mac OS versions; `dext` is for newer versions. To support a wider range of systems that are actively in use, I decided to go for the latter.

<details>
<summary>There are two options to obtain <code>dext</code> (feel free to skip this part):</summary>

1. Install [pqrs-org/Karabiner-Elements](https://github.com/pqrs-org/Karabiner-Elements) since it comes with `dext`.
 
   There is a Cask for Karabiner Elements. I have learned it's not possible to specify a Cask as a dependency of a Formula. Anyway, depending on Karabiner Element has the downside that the version the Formulae builds potentially not to be compatible with the one that comes with Karabiner Elements. (The currently active pull request kmonad/kmonad#780 is an example of KMonad catching up with the version that comes shipped with Karabiner Elements.)

2. Install the PKG for `dext` that comes with KMonad as part of a git submodule

   As outlined in the installation instructions, it's pretty straightforward to install the version of `dext` that is bundled with KMonad in form of a git submodule if you just go for the pre-compiled binary version. You just have to install a PKG file. That said, I suspect this shouldn't  be done as part of the Formulae because Formulaes are not meant for PKGs. There is indeed a way to compile the PKG as part of the Formulae but [that would require signing the extension](https://github.com/pqrs-org/Karabiner-DriverKit-VirtualHIDDevice#steps), which I would like to avoid because it sounds like a rabbit hole.
</details>

While `dext` comes with `karabiner-elements`, given that the version shipped with `kmonad` and `karabiner-elements` can be out of sync, I suggest that I create a separate Cask responsible for installing the version that comes with `kmonad` as a submodule. Thus, I'm suggesting the following set-up:

Create a `kmonad` Formulae that's responsible for building KMonad, and create a `kmonad-driverkit` Cask that installs the PKG that comes with KMonad as a submodule. In the _caveats_ section of the `kmonad` Formulae, highlight that the `kmonad-driverkit` Cask is required and the `karabiner-elements` Cask conflicts with the `kmonad-driverkit` Cask.

**Comment by p-linnane** [comment, marked as answer]
Formulae cannot depend on casks. Formulae in homebrew-core must be both open source and built from source. If something needs system extensions it can only be added as a cask, since those must be signed.

**Reply by aryairani** [reply]
Is it true that formulae in a tap outside of homebrew-core still cannot depend on casks? (see https://github.com/orgs/Homebrew/discussions/5275#discussioncomment-9066648)
